File: js/noise.js
// Simple Perlin noise implementation for terrain generation
class PerlinNoise {
    constructor(seed = 0) {
        this.seed = seed;
        this.permutation = this.generatePermutation();
    }

    generatePermutation() {
        const perm = [];
        for (let i = 0; i < 256; i++) perm[i] = i;
        
        // Shuffle using seed
        for (let i = 255; i > 0; i--) {
            const j = Math.floor(this.seededRandom() * (i + 1));
            [perm[i], perm[j]] = [perm[j], perm[i]];
        }
        
        // Duplicate
        for (let i = 0; i < 256; i++) perm[256 + i] = perm[i];
        return perm;
    }

    seededRandom() {
        this.seed = (this.seed * 16807 + 0) % 2147483647;
        return (this.seed - 1) / 2147483646;
    }

    fade(t) {
        return t * t * t * (t * (t * 6 - 15) + 10);
    }

    lerp(t, a, b) {
        return a + t * (b - a);
    }

    grad(hash, x, y) {
        const h = hash & 3;
        const u = h < 2 ? x : y;
        const v = h < 2 ? y : x;
        return ((h & 1) ? -u : u) + ((h & 2) ? -v : v);
    }

    noise(x, y) {
        const X = Math.floor(x) & 255;
        const Y = Math.floor(y) & 255;
        
        x -= Math.floor(x);
        y -= Math.floor(y);
        
        const u = this.fade(x);
        const v = this.fade(y);
        
        const A = this.permutation[X] + Y;
        const B = this.permutation[X + 1] + Y;
        
        return this.lerp(v,
            this.lerp(u, this.grad(this.permutation[A], x, y),
                         this.grad(this.permutation[B], x - 1, y)),
            this.lerp(u, this.grad(this.permutation[A + 1], x, y - 1),
                         this.grad(this.permutation[B + 1], x - 1, y - 1))
        );
    }

    octaveNoise(x, y, octaves = 4, persistence = 0.5) {
        let total = 0;
        let frequency = 1;
        let amplitude = 1;
        let maxValue = 0;
        
        for (let i = 0; i < octaves; i++) {
            total += this.noise(x * frequency, y * frequency) * amplitude;
            maxValue += amplitude;
            amplitude *= persistence;
            frequency *= 2;
        }
        
        return total / maxValue;
    }
}

window.PerlinNoise = PerlinNoise;

File: server/engine/noise.py
"""Generador de ruido Perlin compatible con js/noise.js."""

import math
from typing import Self


class PerlinNoise:
    """Implementación Python del PerlinNoise de VoxelQuest (js/noise.js).

    Usa el mismo generador congruencial lineal (LCG) y mismas constantes
    para producir terreno idéntico al cliente original.
    """

    def __init__(self, seed: int = 0) -> None:
        self._initial_seed = seed
        self.seed = seed
        self.permutation = self._generate_permutation()

    def _generate_permutation(self) -> list[int]:
        perm = list(range(256))
        for i in range(255, 0, -1):
            j = math.floor(self._seeded_random() * (i + 1))
            perm[i], perm[j] = perm[j], perm[i]
        # Duplicar para evitar buffer overflow al interpolar
        return perm + perm

    def _seeded_random(self) -> float:
        # Mismas constantes que js/noise.js
        self.seed = (self.seed * 16807 + 0) % 2147483647
        return (self.seed - 1) / 2147483646

    @staticmethod
    def _fade(t: float) -> float:
        return t * t * t * (t * (t * 6 - 15) + 10)

    @staticmethod
    def _lerp(t: float, a: float, b: float) -> float:
        return a + t * (b - a)

    @staticmethod
    def _grad(hash_: int, x: float, y: float) -> float:
        h = hash_ & 3
        u = x if h < 2 else y
        v = y if h < 2 else x
        return ((-u if h & 1 else u) + (-v if h & 2 else v))

    def noise(self, x: float, y: float) -> float:
        X = math.floor(x) & 255
        Y = math.floor(y) & 255
        x -= math.floor(x)
        y -= math.floor(y)
        u = self._fade(x)
        v = self._fade(y)

        A = self.permutation[X] + Y
        B = self.permutation[X + 1] + Y

        return self._lerp(
            v,
            self._lerp(
                u,
                self._grad(self.permutation[A], x, y),
                self._grad(self.permutation[B], x - 1, y),
            ),
            self._lerp(
                u,
                self._grad(self.permutation[A + 1], x, y - 1),
                self._grad(self.permutation[B + 1], x - 1, y - 1),
            ),
        )

    def octave_noise(self, x: float, y: float, octaves: int = 4, persistence: float = 0.5) -> float:
        total = 0.0
        frequency = 1.0
        amplitude = 1.0
        max_value = 0.0

        for _ in range(octaves):
            total += self.noise(x * frequency, y * frequency) * amplitude
            max_value += amplitude
            amplitude *= persistence
            frequency *= 2

        return total / max_value

    def with_seed(self, seed: int) -> Self:
        """Devuelve una nueva instancia con otra seed (útil para tests)."""
        return self.__class__(seed)
